### PR TITLE
[Performance] Split FlexAttention and FlashInfer attention and cache update

### PR DESCRIPTION
**Pull request by Etelis**
This PR separates the KV-cache update from the attention forward pass for the FlexAttention and FlashInfer backends, following the pattern established in #25954 for FlashAttention.

Addresses #32335.

### Changes

**FlexAttention Backend (`vllm/v1/attention/backends/flex_attention.py`)**
- Set `forward_includes_kv_cache_update = False` on `FlexAttentionBackend`
- Implemented `do_kv_cache_update()` method that calls `reshape_and_cache_flash` to store keys/values in cache
- Removed KV cache update logic from `forward()` method

**FlashInfer Backend (`vllm/v1/attention/backends/flashinfer.py`)**
- Set `forward_includes_kv_cache_update = False` on `FlashInferBackend`
- Implemented `do_kv_cache_update()` method with FlashInfer-specific cache layout handling
- Removed KV cache update logic from `forward()` method

**Base Class (`vllm/v1/attention/backend.py`)**
- Added default `do_kv_cache_update()` implementation to `AttentionImpl`
- Added `slot_mapping` attribute to `AttentionMetadata`

## Related PRs

- #25954 - Original implementation for FlashAttention backend
- #32335 - Tracking issue for all backend implementations

---

## Benchmark Results

**Model**: Qwen/Qwen3-1.7B | **Backend**: FLASH_ATTN

### Latency (batch_size=8, input=512, output=128)

| Branch | Avg (s) | P50 (s) | P99 (s) |
|--------|---------|---------|---------| 
| main | 1.0054 | 1.0025 | 1.0240 |
| PR | 1.0059 | 1.0056 | 1.0107 |

### Throughput (100 prompts)

| Branch | Requests/s | Tokens/s |
|--------|------------|----------|
| main | 12.61 | 14,525 |
| PR | 12.59 | 14,499 |

### CODE REVIEW

File: vllm/v1/attention/backends/rocm_aiter_fa.py [hunk 1]
```diff
@@ -754,6 +756,68 @@ def __init__(
                 "Encoder self-attention is not implemented for FlashAttentionImpl"
             )
 
+    def do_kv_cache_update(
```
Comment (tjtanaa) [inline vllm/v1/attention/backends/rocm_aiter_fa.py]: theorectically should be fine. Let me validate this locally as this preshuffle setting is not covered in the CI.

File: vllm/attention/layer.py [hunk 2]
```diff
@@ -834,7 +834,11 @@ def unified_kv_cache_update(
         f"Expected slot_mapping to be a dict, got {type(slot_mapping)}. "
     )
     layer_slot_mapping = slot_mapping.get(layer_name)
-    if layer_slot_mapping is not None:
+
+    # Skip KV cache update if key/value is None or slot_mapping is not available.
+    # key and value may be None in the case of cross attention where they are
+    # calculated once based on the output from the encoder and then cached.
+    if layer_slot_mapping is not None and key is not None and value is not None:
```
Comment (ProExpertProg) [inline vllm/attention/layer.py]: Can we do this even earlier, inside the actual layer class before calling `torch.ops.vllm.unfied_kv_cache_update`?

File: vllm/v1/attention/backend.py [hunk 3]
```diff
@@ -693,6 +702,64 @@ def forward(
...
+        """
+        # Skip if sharing KV cache with an earlier attention layer
+        if self.kv_sharing_target_layer_name is not None:
+            return kv_cache
+
+        key_cache, value_cache = kv_cache.unbind(0)
+
+        # Extract slot_mapping from metadata if it's an object, otherwise use directly
+        slot_mapping = (
+            attn_metadata.slot_mapping
```
Comment (ProExpertProg) [inline vllm/v1/attention/backend.py]: Let's always use the slot mapping directly if possible - are there backends that do something special here?

File: vllm/v1/attention/backends/flashinfer.py [hunk 4]
```diff
@@ -1200,6 +1201,60 @@ def process_weights_after_loading(self, act_dtype: torch.dtype):
...
+
+        # The FlashInfer API requires data to be in fp8_e4m3 or fp8_e5m2
+        # to process the cache when the kv_cache_dtype is fp8
+        if self.kv_cache_dtype.startswith("fp8"):
+            torch_dtype = FlashInferBackend.get_fp8_dtype_for_flashinfer(
+                self.kv_cache_dtype
+            )
+            kv_cache = kv_cache.view(torch_dtype)
+
+        return kv_cache
```
Comment (ProExpertProg) [inline vllm/v1/attention/backends/flashinfer.py]: Why returning? This feels like something that should remain at the start of forward and we can just defer to the super method. I also don't think this return goes anywhere

File: vllm/v1/attention/backends/flashinfer.py [hunk 5]
```diff
@@ -1192,6 +1193,55 @@ def process_weights_after_loading(self, act_dtype: torch.dtype):
         if self.sinks is not None and self.sinks.dtype != torch.float32:
             self.sinks = self.sinks.to(torch.float32)
 
+    def do_kv_cache_update(
+        self,
+        layer: torch.nn.Module,
```
[the remark this answers is not in the document]
Comment (ProExpertProg) [reply vllm/v1/attention/backends/flashinfer.py]: I think I agree here

[the remark this answers is not in the document]
Comment (tjtanaa) [reply vllm/v1/attention/backends/flashinfer.py]: @ProExpertProg should we do this for all of the backends? The PRs are just following the initial PR that split FlashAttn attention and cache update https://github.com/vllm-project/vllm/pull/25954 . The initial PR is using `torch.nn.Module`

[the remark this answers is not in the document]
Comment (tjtanaa) [reply vllm/v1/attention/backends/flashinfer.py]: e.g this PR https://github.com/vllm-project/vllm/pull/33106 is also following the initial PR which uses `torch.nn.Module`.

[the remark this answers is not in the document]
Comment (ProExpertProg) [reply vllm/v1/attention/backends/flashinfer.py]: Yeah I think so, I didn't notice that in the first PR

[the remark this answers is not in the document]
Comment (tjtanaa) [reply vllm/v1/attention/backends/flashinfer.py]: Ok. Then I will help to comment on the PRs so that they use `AttentionLayer`.

File: vllm/v1/attention/backends/rocm_aiter_fa.py (hunk 1 above)
Comment (tjtanaa) [reply vllm/v1/attention/backends/rocm_aiter_fa.py]: Validated the rocm fa preshuffle feature, the accuracy is fine.

**Comment by Etelis** [comment]
> @Etelis Do the `lm_eval` runs also have to be redone?

Yes they're working :)

File: vllm/v1/attention/backend.py (hunk 3 above)
Comment (ElizaWszola) [reply vllm/v1/attention/backend.py]: +1, I think directly passing slot mapping as a function arg would be better for readibility

File: vllm/v1/attention/backends/rocm_aiter_fa.py [hunk 6]
```diff
@@ -754,6 +756,68 @@ def __init__(
...
+        self,
+        layer: torch.nn.Module,
```
[the remark this answers is not in the document]
Comment (tjtanaa) [reply vllm/v1/attention/backends/rocm_aiter_fa.py]: @Etelis Please apply this suggestion.

File: vllm/v1/attention/backends/flashinfer.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (tjtanaa) [reply vllm/v1/attention/backends/flashinfer.py]: @Etelis let's update to AttentionLayer

File: vllm/v1/attention/backends/rocm_aiter_fa.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (Etelis) [reply vllm/v1/attention/backends/rocm_aiter_fa.py]: Done! Updated to `AttentionLayer` in commit eb611b152.

File: vllm/v1/attention/backends/flashinfer.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Etelis) [reply vllm/v1/attention/backends/flashinfer.py]: Done! Updated to `AttentionLayer` in commit eb611b152.

File: vllm/v1/attention/backend.py (hunk 3 above)
Comment (Etelis) [reply vllm/v1/attention/backend.py]: FlashInfer and RocmAiterFA need `attn_metadata` because they access other fields beyond slot_mapping (e.g., `attn_metadata.k_scale`, `attn_metadata.v_scale` in the shuffle case). The base class signature uses `attn_metadata: T` which is the generic pattern.

File: vllm/v1/attention/backends/flashinfer.py (hunk 4 above)
Comment (Etelis) [reply vllm/v1/attention/backends/flashinfer.py]: Fixed in ac5d571 and 1967079.

File: vllm/attention/layer.py (hunk 2 above)
Comment (EtelisIBM) [reply vllm/attention/layer.py]: The check is now in do_kv_cache_update which is called separately from forward. This keeps the logic contained within the cache update method.

**Comment by tjtanaa** [comment]
@Etelis Let me evaluate on AMD the backend. There is this PR https://github.com/vllm-project/vllm/pull/33463 that says that it is not ready for the split.

File: vllm/model_executor/layers/quantization/utils/fp8_utils.py [hunk 7]
```diff
@@ -426,6 +426,24 @@ def _run_deepgemm(
         weight: torch.Tensor,
         weight_scale: torch.Tensor,
     ) -> torch.Tensor:
+        # Pad K dimension to multiple of block size if needed.
```
Comment (ProExpertProg) [inline vllm/model_executor/layers/quantization/utils/fp8_utils.py]: Is this a bad merge?

File: vllm/model_executor/layers/quantization/utils/fp8_utils.py [hunk 8]
```diff
@@ -547,6 +565,22 @@ def _run_flashinfer(
         This backend uses TensorRT-LLM's FP8 block-scale GEMM kernels
         and supports FP8+FP8 (W8A8 full quantization) on SM90+ (Hopper).
         """
+        # Pad K dimension to multiple of block size if needed.
```
Comment (ProExpertProg) [inline vllm/model_executor/layers/quantization/utils/fp8_utils.py]: same here

File: vllm/v1/attention/backends/flashinfer.py [hunk 9]
```diff
@@ -1200,6 +1202,46 @@ def process_weights_after_loading(self, act_dtype: torch.dtype):
...
+        # Skip if sharing KV cache with an earlier attention layer
+        if self.kv_sharing_target_layer_name is not None:
+            return
+
+        # NOTE(woosuk): Here, key and value are padded while slot_mapping is
+        # not padded. However, we don't need to do key[:num_actual_tokens]
+        # and value[:num_actual_tokens] because the reshape_and_cache_flash
+        # op uses the slot_mapping's shape to determine the number of
+        # actual tokens.
+        torch.ops._C_cache_ops.reshape_and_cache_flash(
```
Comment (ProExpertProg) [inline vllm/v1/attention/backends/flashinfer.py:1234]: Use `super()`?

File: vllm/v1/attention/backends/flashinfer.py [hunk 10]
```diff
@@ -1325,6 +1344,14 @@ def forward(
         output_padded = output
         output = output[:num_actual_tokens]
 
+        # The FlashInfer API requires data to be in fp8_e4m3 or fp8_e5m2
+        # to process the cache when the kv_cache_dtype is fp8.
+        if self.kv_cache_dtype.startswith("fp8"):
```
Comment (ProExpertProg) [inline vllm/v1/attention/backends/flashinfer.py:1349]: Missing the `if self.kv_sharing_target_layer_name is None` from before?

File: vllm/v1/attention/backend.py [hunk 11]
```diff
@@ -600,6 +600,15 @@ class AttentionImpl(ABC, Generic[T]):
...
     scale: float
 
+    # KV cache configuration - set by subclasses in __init__
+    kv_cache_dtype: str
+    kv_sharing_target_layer_name: str | int | None
+
+    # Whether the forward method includes KV cache update.
+    # When False, the KV cache update is handled separately via do_kv_cache_update().
+    # This allows for better performance optimizations and independent scheduling.
+    forward_includes_kv_cache: bool = True
```
Comment (ProExpertProg) [inline vllm/v1/attention/backend.py]: Remove this: `forward_includes_kv_cache_update` already exists on the backend?

**Review by ProExpertProg — CHANGES_REQUESTED** [review]

File: vllm/v1/attention/backend.py [hunk 12]
```diff
@@ -277,7 +277,7 @@ def get_required_kv_cache_layout(cls) -> "KVCacheLayoutType | None":
 
 
 class AttentionMetadata:
-    pass
+    slot_mapping: torch.Tensor
```
Comment (ProExpertProg) [inline vllm/v1/attention/backend.py:280]: Why is this needed? Should we not be reading the slot mapping from the forward context directly?

File: vllm/v1/attention/backend.py [hunk 13]
```diff
@@ -707,6 +711,54 @@ def forward(
...
+            attn_metadata: Attention metadata containing slot_mapping, or
+                           slot_mapping tensor directly.
+        """
+        # Skip if sharing KV cache with an earlier attention layer
+        if self.kv_sharing_target_layer_name is not None:
+            return
+
+        key_cache, value_cache = kv_cache.unbind(0)
+
+        slot_mapping = attn_metadata.slot_mapping
```
Comment (ProExpertProg) [inline vllm/v1/attention/backend.py]: Why not always use slot mapping directly? It should come from the forward context slot mapping and not attention metadata

File: vllm/v1/attention/backend.py (hunk 12 above)
Comment (Etelis) [reply vllm/v1/attention/backend.py:280]: Lines 280, 743 in backend.py:
Changed do_kv_cache_update to accept slot_mapping: torch.Tensor directly instead of extracting it from attn_metadata. This matches how the caller (unified_kv_cache_update in attention.py) already gets slot_mapping from the forward context and passes it directly.

FlashInfer (line 1234):
Updated to use the slot_mapping parameter directly. Couldn't use super() here because FlashInfer's cache layout is [num_blocks, 2, block_size, num_kv_heads, head_size] and it accesses with kv_cache[:, 0] and kv_cache[:, 1], while the base class uses kv_cache.unbind(0) expecting [2, ...] as the first dimension.

File: vllm/model_executor/layers/attention/attention.py [hunk 14]
```diff
@@ -637,6 +640,9 @@ def unified_kv_cache_update(
         f"Expected slot_mapping to be a dict, got {type(slot_mapping)}. "
     )
     layer_slot_mapping = slot_mapping.get(layer_name)
+
+    # Skip KV cache update if slot_mapping is not available for this layer.
+    # Note: key and value are guaranteed to be non-None by the caller.
```
Comment (ElizaWszola) [inline vllm/model_executor/layers/attention/attention.py:645]: nit: can you add an assert for this?
